cpp/src/arrow/dataset/dataset_encryption_test.cc

```diff
@@ -0,0 +1,402 @@
+                                    "part=i/part0.parquet", "part=j/part0.parquet"};
+  ValidateFilesExist(mock_fs, files);
+
+  // ----- Read the Dataset -----
+
+  // Get FileInfo objects for all files under the base directory
+  arrow::fs::FileSelector selector;
+  selector.base_dir = kBaseDir;
+  selector.recursive = true;
+
+  // Create a FileSystemDatasetFactory
+  arrow::dataset::FileSystemFactoryOptions factory_options;
+  factory_options.partitioning = partitioning;
+  factory_options.partition_base_dir = kBaseDir;
+  ASSERT_OK_AND_ASSIGN(auto dataset_factory,
+                       arrow::dataset::FileSystemDatasetFactory::Make(
+                           mock_fs, selector, file_format, factory_options));
+  // Create a Dataset
+  ASSERT_OK_AND_ASSIGN(auto dataset_in, dataset_factory->Finish());
+
+  // Define the callback function
+  std::function<arrow::Status(arrow::dataset::TaggedRecordBatch tagged_record_batch)>
+      visitor =
+          [](arrow::dataset::TaggedRecordBatch tagged_record_batch) -> arrow::Status {
+    return arrow::Status::OK();
+  };
+
+  ASSERT_OK_AND_ASSIGN(auto scanner_builder_in, dataset_in->NewScan());
+  ASSERT_OK_AND_ASSIGN(auto scanner_in, scanner_builder_in->Finish());
+
+  // Scan the dataset and process the record batches using the callback function
+  arrow::Status status = scanner_in->Scan(visitor);
+
+  // Check if there was an error during iteration
+  ASSERT_OK(status);
+}
+
+// Write dataset to disk with encryption and then read in a single parquet file
+TEST_F(DatasetEncryptionTest, WriteReadSingleFile) {
+  auto [dataset_encryption_config, dataset_decryption_config] =
+      CreateDatasetEncryptionConfig(kColumnMasterKeysIds, kColumnMasterKeys, kNumColumns,
+                                    kFooterKeyMasterKeyId, kFooterKeyMasterKey);
+
+  auto parquet_scan_options = std::make_shared<ParquetFragmentScanOptions>();
+  parquet_scan_options->SetDatasetDecryptionConfig(dataset_decryption_config);
+
+  // create our Parquet file format object
+  auto file_format = std::make_shared<ParquetFileFormat>();
+  // update default scan options
+  file_format->default_fragment_scan_options = parquet_scan_options;
+
+  // set write options
+  auto file_write_options = file_format->DefaultWriteOptions();
+  std::shared_ptr<ParquetFileWriteOptions> parquet_file_write_options =
+      std::static_pointer_cast<ParquetFileWriteOptions>(file_write_options);
+  parquet_file_write_options->SetDatasetEncryptionConfig(dataset_encryption_config);
+
+  // create our mock file system
+  ::arrow::fs::TimePoint mock_now = std::chrono::system_clock::now();
+  ASSERT_OK_AND_ASSIGN(auto file_system,
+                       ::arrow::fs::internal::MockFileSystem::Make(mock_now, {}));
+  // create filesystem
+  ASSERT_OK(file_system->CreateDir(""));
+
+  auto mock_fs =
+      std::dynamic_pointer_cast<::arrow::fs::internal::MockFileSystem>(file_system);
```

Comment:
```suggestion
      internal::checked_pointer_cast<::arrow::fs::internal::MockFileSystem>(file_system);
```

Reply:
BTW, do we really need this cast? It seems that you only need `mock_fs->OpenInputFile(file_path)` and that's already available in the base FileSystem class.